### ISSUE TITLE
Server-client communication (suggested logic - feedback please!)

### ISSUE THREAD

**Issue by emilyb7** [CYPIAPT-LNDSE/goals-app#137]
We can use web sockets for transferring data between the front- and backend. 

For this we can use the `socket.io` library for the client and the server. See #128. 

Suggested steps:

1. Refactor state so that goals are spread across multiple arrays:
- `goals_existing` (goals that are loaded from the database when user logs in, or which have been added to the DB in the meantime)
- `goals_added` - goals the user has added that have not yet been processed server-side
- `goals_edited` - goals the user has made changes to that have not yet been processes server-side
- `goals_deleted`goals the user has deleted client-side, but which are not yet set to deleted in the DB

Goals deleted would not be shown to the user.

2. Process changes on page re-render
Each time the page re-renders, we check whether any of added, edited or deleted arrays contain anything. If so, we send a message to the server via the socket. We would probably need to add a status property to the goal, so that we make sure the events have been fired only once.

3. Server-side
Server receives message, and makes changes to DB. On successful processing of the changes, send a message back to the client.
- goals added can be added to the existing goals array
- goals deleted can be cleared on client side
- goals edited can be moved back to existing goals array

4. Server-side errors / network connection fails
Need to inform the user that their changes have not been processed. 
If no network connection, check for restored connection and try again.

5. Stretch goal
Store in local storage until connection comes back

6. Channels (if not too tricky)
We can use channels to handle the case that a user is signed in on multiple devices. Create a channel for all instances of the user being signed in. If a user makes changes on one device, ping other clients and send the user a message telling them to refresh the page.

7. Saving ratings
Implement similar logic (multiple arrays) for new ratings

**Comment by emilyb7** [comment]
Pros of web sockets:
- check for sessions on other devices / in other browsers
- could integrate will with PWA (offline) functionality in future iterations

Cons of web sockets:
- not an orthodox solution, AJAX is a more common tool
- may make server-side architecture more complex
- need to work out how to combine use of sockets with user authentication

**Comment by emilyb7** [comment]
Resources on sockets + auth:

https://devcenter.heroku.com/articles/websocket-security

**Comment by des-des** [comment]
@emilyb7 maybe add a pic of some of that stuff we drew out, so it get documented
